src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1096 @@
+	worker.bgw_main_arg = ObjectIdGetDatum(MyDatabaseId);
+	memcpy_s(worker.bgw_extra, sizeof(worker.bgw_extra), &extensionOwner,
+			 sizeof(Oid));
+	worker.bgw_notify_pid = MyProcPid;
+
+	if (!RegisterDynamicBackgroundWorker(&worker, &handle))
+	{
+		return NULL;
+	}
+
+	pid_t pid;
+	WaitForBackgroundWorkerStartup(handle, &pid);
+
+	return handle;
+}
+
+
+void
+CitusBackgroundTaskQueueMonitorMain(Datum arg)
+{
+	Oid databaseOid = DatumGetObjectId(arg);
+
+	/* extension owner is passed via bgw_extra */
+	Oid extensionOwner = InvalidOid;
+	memcpy_s(&extensionOwner, sizeof(extensionOwner),
+			 MyBgworkerEntry->bgw_extra, sizeof(Oid));
+
+	BackgroundWorkerUnblockSignals();
+
+	/* connect to database, after that we can actually access catalogs */
+	BackgroundWorkerInitializeConnectionByOid(databaseOid, extensionOwner, 0);
+
+	/*
+	 * There should be exactly one background task monitor running, running multiple would
+	 * cause conflicts on processing the tasks in the catalog table as well as violate
+	 * parallelism guarantees. To make sure there is at most, exactly one backend running
+	 * we take a session lock on the CITUS_BACKGROUND_TASK_MONITOR operation.
+	 *
+	 * TODO now that we have a lock, we should install a term handler to terminate any
+	 * 'child' backend when we are terminated. Otherwise we will still have a situation
+	 * where the actual task could be running multiple times.
+	 */
+	LOCKTAG tag = { 0 };
+	SET_LOCKTAG_CITUS_OPERATION(tag, CITUS_BACKGROUND_TASK_MONITOR);
+	const bool sessionLock = true;
+	const bool dontWait = true;
+	LockAcquireResult locked =
+		LockAcquire(&tag, AccessExclusiveLock, sessionLock, dontWait);
+	if (locked == LOCKACQUIRE_NOT_AVAIL)
+	{
+		ereport(ERROR, (errmsg("background task queue monitor already running for "
+							   "database")));
+		exit(0);
+	}
+
+	/* make worker recognizable in pg_stat_activity */
+	pgstat_report_appname("citus background task queue monitor");
+
+	ereport(DEBUG1, (errmsg("started citus background task queue monitor")));
+
+	/* TODO this is here for debugging purposses, remove before merge. */
+	if (BackgroundTaskMonitorDebugDelay)
+	{
+		pg_usleep(30 * 1000 * 1000);
+	}
+
+	MemoryContext perTaskContext = AllocSetContextCreate(CurrentMemoryContext,
+														 "PerTaskContext",
+														 ALLOCSET_DEFAULT_MINSIZE,
+														 ALLOCSET_DEFAULT_INITSIZE,
+														 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/*
+	 * First we find all jobs that are running, we need to check if they are still running
+	 * if not reset their state back to scheduled.
+	 */
+	{
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		ResetRunningBackgroundTasks();
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+	}
+
+
+	MemoryContext oldContextPerJob = MemoryContextSwitchTo(perTaskContext);
+	bool hasJobs = true;
+	while (hasJobs)
+	{
+		MemoryContextReset(perTaskContext);
+
+		CHECK_FOR_INTERRUPTS();
+
+		InvalidateMetadataSystemCache();
```

Comment:
do we use the metadata cache here?

Reply:
The call does two things:
 - clear connections - we don't use those
 - clears `MetadataCacheData` which contains the Oid's of our relations/indices/enums/etc - we do use those

Since we use the cached citus object Oid's I guess the answer is yes. Although I don't directly see how the Oid's could become invalid between iterations.

I think I saw it happening in other background workers when looping, checked that it cleared something we use here, and copied.